server/src/main/java/org/elasticsearch/index/mapper/DataStreamTimestampFieldMapper.java

```diff
@@ -0,0 +1,189 @@
+
+        @Override
+        public MetadataFieldMapper build(BuilderContext context) {
+            return new DataStreamTimestampFieldMapper(
+                fieldType,
+                new DataStreamTimestampFieldFieldType(),
+                fieldName
+            );
+        }
+    }
+
+    public static class TypeParser implements MetadataFieldMapper.TypeParser {
+
+        @Override
+        public MetadataFieldMapper.Builder<?> parse(String name,
+                                                    Map<String, Object> node,
+                                                    ParserContext parserContext) throws MapperParsingException {
+            Builder builder = new Builder();
+            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext();) {
+                Map.Entry<String, Object> entry = iterator.next();
+                String fieldName = entry.getKey();
+                Object fieldNode = entry.getValue();
+                if (fieldName.equals("field_name")) {
+                    builder.setFieldName((String) fieldNode);
+                    iterator.remove();
+                }
+            }
+            return builder;
+        }
+
+        @Override
+        public MetadataFieldMapper getDefault(ParserContext parserContext) {
+            return new DataStreamTimestampFieldMapper(Defaults.DATA_STREAM_TIMESTAMP_FIELD_TYPE,
+                new DataStreamTimestampFieldFieldType(), null);
+        }
+    }
+
+    private final String fieldName;
+
+    private DataStreamTimestampFieldMapper(FieldType fieldType, MappedFieldType mappedFieldType, String fieldName) {
+        super(fieldType, mappedFieldType);
+        this.fieldName = fieldName;
+    }
+
+    @Override
+    public void preParse(ParseContext context) throws IOException {
+
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
+
+    }
+
+    @Override
+    public void postParse(ParseContext context) throws IOException {
+        if (fieldName == null) {
+            // not configured, so skip the validation
+            return;
+        }
+
+        IndexableField[] fields = context.rootDoc().getFields(fieldName);
+        if (fields.length == 0) {
+            throw new IllegalArgumentException("data stream timestamp field [" + fieldName + "] is missing");
+        }
+
+        long numberOfValues =
+            Arrays.stream(fields).filter(indexableField -> indexableField.fieldType().pointDimensionCount() > 0).count();
+        if (numberOfValues > 1) {
+            throw new IllegalArgumentException("data stream timestamp field [" + fieldName + "] encountered multiple values");
+        }
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        boolean includeDefaults = params.paramAsBoolean("include_defaults", false);
+        if (includeDefaults == false && fieldName == null) {
```

Comment:
shouldn't we always return when the fieldName is null ?

Reply:
I'm ok with doing that, I just thought that if `includeDefaults` is `true` then that the fieldName should be returned anyway, irregardless whether it has a value. But I guess it is ok to ignore `includeDefaults` if `fieldName` is null?